feat(queryByCurrentValue)

eunjae-lee opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: http://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**: It adds `queryByCurrentValue`, `getByCurrentValue`, `queryAllByCurrentValue` and `getAllByCurrentValue`.

<!-- Why are these changes necessary? -->

**Why**: `getByValue` cannot get elements with `value` property (it only checks DOM `value` attributes)
(Discussed at #166)

<!-- How were these changes implemented? -->

**How**: Added a set of `(get|query)(All)?ByCurrentValue` methods.

<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [ ] Documentation
- [x] Tests
- [ ] Ready to be merged <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->
- [ ] Added myself to contributors table <!-- this is optional, see the contributing guidelines for instructions -->

<!-- feel free to add additional comments -->

Feel free to give any feedbacks.

---

Closes #158

kentcdodds commented on src/queries.js:
```diff
@@ -356,8 +387,12 @@ export {
   queryByValue,
   queryAllByValue,
+  queryByCurrentValue,
+  queryAllByCurrentValue,
```
Could you move both of these down two lines. I kinda like to group things together.

kentcdodds commented on src/queries.js:
```diff
@@ -180,6 +180,22 @@ function queryByAltText(...args) {
+  const matcher = exact ? matches : fuzzyMatches
+  const matchOpts = {collapseWhitespace, trim}
+  return Array.from(container.querySelectorAll(`input,textarea,select`)).filter(
+    node => matcher(node.value, node, value, matchOpts),
```
This wont work for selects. Please look at (and maybe just use) the logic in `queryAllBySelectText` for selects.

kentcdodds commented on src/__tests__/element-queries.js:
```diff
@@ -565,4 +565,19 @@ test('getByText ignores script tags by default', () => {
   expect(getAllByText(/hello/i, {ignore: false})).toHaveLength(3)
 })
 
+test('get element by its dynamically assigned value', () => {
```
Could we also have a test for a textarea and a select. Can be in this test or in a new one I don't care which.

eunjae-lee commented:
Thanks for the review. I made changes regarding your three comments.

alexkrolick commented:
Looks good! Just need docs and a test to execute line 358 (the error state for queryAllBy...) to keep coverage at 100%.

Also it might be nice to add an integration test for `fireEvent.change()` with `getCurrentValue`.

kentcdodds commented on src/__tests__/element-queries.js:
```diff
@@ -565,4 +565,63 @@ test('getByText ignores script tags by default', () => {
+  expect(queryByCurrentValue('Alaska').id).toEqual('state-select')
+
+  getByTestId('state').value = 'AL'
+  expect(getByCurrentValue('Alabama').id).toEqual('state-select')
```
I wonder if it's at all confusing that we are expected to set the `value` to `AL` but when we use the query we search for `Alabama`.

I realize that's just how things work (the value for a select has a display value), but I feel like this could be a point of confusion for people. My guess is they'd prefer to select by `AL` instead which isn't what is shown to the user but feels more intuitive here. I'm not sure of the best solution to this... Maybe allow for either the display value or the actual value?

eunjae-lee replied:
That makes sense. To end-users `Alabama` is a value, but `AL` is the actual value.

eunjae-lee replied:
However, `selectElem.value` could be either `1`, `AL`, or whatever user cannot know. In a perspective of writing tests as user would use it, I'd rather choose to restrict it to display value.
What do you think?

kentcdodds replied:
Yeah, I think `displayValue` is probably a better mental model, though it may be a tiny confusing for `input` and `textarea` where the _display_ value is actually the _actual_ value but I think that's alright. Besides, people should normally be getting these elements by their label anyway, so I'm not too concerned about this slightly less-awesome API.

Let's go with `getByDisplayValue` 👍

eunjae-lee commented:
@alexkrolick I've added a test to check if it throws.

@kentcdodds I've renamed and updates the README.md. I didn't know how to edit it, so it's kind of a draft. Where do you want to put the section in the document? And how do you want to explain it to users? Take a look at the doc and let me know. Besides, I'm not a native English speaker, so pretty much worried about it 😅

kentcdodds commented on README.md:
```diff
@@ -86,6 +86,7 @@ when a real user uses it.
   - [`getByAltText`](#getbyalttext)
   - [`getByTitle`](#getbytitle)
   - [`getByValue`](#getbyvalue)
```
Let's remove the docs for `getByValue` and `getBySelectText` because we want to encourage people to use the `getByDisplayValue` instead.

kentcdodds commented on README.md:
````diff
@@ -409,6 +410,55 @@ Returns the element that has the matching value.
+  }): HTMLElement
+```
+
+Returns the element that has the matching display value.
````
> Returns the `input`, `textarea`, or `select` element that has the matching display value.

kentcdodds commented on README.md:
```diff
@@ -409,6 +410,55 @@ Returns the element that has the matching value.
+// <input type="text" id="lastName" />
+// document.getElementById('lastName').value = 'Norris'
+
+const lastNameInput = getByDisplayValue('Norris')
```
> const lastNameInput = getByDisplayValue(document.body, 'Norris')

kentcdodds commented on README.md:
```diff
@@ -409,6 +410,55 @@ Returns the element that has the matching value.
+// <textarea id="messageTextArea"></textarea>
+// document.getElementById('messageTextArea').value = 'Hello World'
+
+const messageTextArea = getByDisplayValue('Hello World')
```
> const messageTextArea = getByDisplayValue(document.body, 'Hello World')

kentcdodds commented on README.md:
```diff
@@ -409,6 +410,55 @@ Returns the element that has the matching value.
+// </select>
+
+const selectElement = getByDisplayName('Alaska')
+console.log(selectElement.id)
```
Remove the `console.log` here (it's not necessary and it's nice to be consistent with the others). Then fix it to be:

> const selectElement = getByDisplayName(document.body, 'Alaska')

eunjae-lee commented:
@kentcdodds done! I've used `container` instead of `document.body` since I found out other parts of the doc used `container`.

kentcdodds commented:
Looks like we're missing some coverage. You can open the coverage report in the browser in `coverage/lcov-report` to see what's missing.

eunjae-lee commented:
@kentcdodds coverage 100% done

alexkrolick commented on README.md:
```diff
@@ -337,77 +335,54 @@ content is in an inline script file, then the script tag could be returned.
 
 If you'd rather disable this behavior, set `ignore` to `false`.
 
-### `getByAltText`
```
Why are these doc sections being deleted?

eunjae-lee replied:
@alexkrolick refer to https://github.com/kentcdodds/dom-testing-library/pull/169#discussion_r240464612

eunjae-lee commented:
Updated the doc. Thanks for pointing it out @alexkrolick

kentcdodds commented:
:tada: This PR is included in version 3.14.0 :tada:

The release is available on:
- `npm package (@latest dist-tag)`
- [GitHub release](https://github.com/kentcdodds/dom-testing-library/releases/tag/v3.14.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

eunjae-lee commented:
Awesome! Thank you :) This merge made my day!!!
